Add power/exponent to math operations

DhruvaG2000 opened the pull request:
<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [X] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [X] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [x] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed) see https://github.com/rizinorg/book/pull/87

**Detailed description**
Can use ``**`` operator for finding the power of an integer


**Test plan**
To implement, 
```c
[0x00000000]> ? 2**3
int32   8
uint32  8
hex     0x8
octal   010
unit    8
segment 0000:0008
string  "\b"
fvalue  8.0
float   0.000000f
double  0.000000
binary  0b00001000
trits   0t22
```

**Closing issues**
 Closes https://github.com/rizinorg/rizin/issues/613


Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

DhruvaG2000 committed:
Add power/exponent to math operations

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

DhruvaG2000 commented:
Facing 2 difficulties which I will try to tackle:
- ~unable to use ``e`` or ``E`` or  ``**``. Not sure why only ``!`` works~ `**` is the new power Operator.
- Will not work for decimal point powers like ``0.5`` or ``1.5`` etc. Not sure how this can be done. Should I include math header and use the ``pow()`` simply?

Rot127 commented:
Also please add some tests in `test/db/cmd/cmd_help`

Rot127 commented:
> unable to use e or E or **. Not sure why only ! works

Letters get treated as `0` for me. So they are parsed somehow.
```
[0x00000000]> ? e
int32   0
uint32  0
...
```
Maybe follow the code path from the `?` command handler and check where letters get replaced or skipped.

DhruvaG2000 committed:
Add tests for power operation

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

DhruvaG2000 committed:
Got ** to work as power operator

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

DhruvaG2000 committed:
Update docs: Mention ** as pow operator

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

Rot127 commented:
Tested with https://github.com/rizinorg/rizin/pull/2220/commits/c7102afc256eb849517e8e1aa475bb1d8ed801b8

```bash
[0x00000000]> ? 2!0
int32   2
# 2!0 != 1

[0x00000000]> ? 2!-1
# freezes. Probably an integer underflow and loops 4billion times :)

[0x00000000]> ? 2!1.4
int32   2 
# If you do not want to support floating point numbers yet, you can print a warning with `RZ_LOG_WARN()`
```

Please also add test cases once you fixed them.

wargio reviewed (changes requested):
Please add tests also for negative (error, no need to implement it) and zero case (needed).

ret2libc commented on librz/include/rz_util/rz_num.h:
```diff
@@ -23,6 +23,7 @@ typedef enum {
 	RNCPLUS = '+',
 	RNCMINUS = '-',
 	RNCMUL = '*',
+	RNCEXP = '**',
```
I think it is not really needed to put this here with any `=... `. In particular, this seems wrong as you can just assign one character to the enum value.

DhruvaG2000 replied:
Yes, true. Then I am changing this to capital `E` is that okay?

DhruvaG2000 committed:
Revert "Update docs: Mention ** as pow operator"

This reverts commit 32b4d180c195809be349b32c5e682e177a621857.

DhruvaG2000 committed:
** operator improvement

- Now throws error for negative exponents
- Returns 1 for power 0

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

DhruvaG2000 committed:
RzNumCalcToken:assign 'E' to the enum value- pow

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

DhruvaG2000 committed:
Pow operator: improvements in implementation

- now supports negative powers
- throws a warning when decimal point power is entered

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

DhruvaG2000 committed:
Add tests to cover all cases for the ** operator

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

DhruvaG2000 committed:
Update docs: Mention ** as pow operator

This reverts commit 82540ee8fb1aac4dc416ba6699dc80463db539cf.

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

ret2libc commented on librz/util/calc.c:
```diff
@@ -379,6 +397,11 @@ static RzNumCalcToken get_token(RzNum *num, RzNumCalc *nc) {
+			return nc->curr_tok = RNCEXP;
+		}
+		cin_putback(num, nc, c);
+		return nc->curr_tok = (RzNumCalcToken)ch;
```
```suggestion
		return nc->curr_tok = RNCMUL;
```

ret2libc commented on librz/util/calc.c:
```diff
@@ -186,6 +186,23 @@ static RzNumCalcValue term(RzNum *num, RzNumCalc *nc, int get) {
+			if (d.d - (int)d.d) {
+				RZ_LOG_WARN("floating point powers not yet supported")
+			}
+			if ((int)d.d > 0) {
+				for (int i = 1; i < (int)d.d; i++) {
+					left = Nmul(exp_left, left);
+				}
+			} else if ((int)d.d < 0) {
+				for (int i = -1; i >= (int)d.d; i--) {
+					left = Ndiv(left, exp_left);
+				}
+			} else {
+				left = Ndiv(left, exp_left);
+			}
```
The logic code should go in a `Nexp` function, similarly to how it's done for all the other operations.

ret2libc commented on librz/util/calc.c:
```diff
@@ -186,6 +186,23 @@ static RzNumCalcValue term(RzNum *num, RzNumCalc *nc, int get) {
 				return d;
 			}
 			left = Ndiv(left, d);
+		} else if (nc->curr_tok == RNCEXP) {
```
Whether to place this here or in `expr` or in `term` depends on which precedence we want to give to the exponential operation.

Could you put some tests with more complicated expressions so we can check it works as we'd want?

In python, the following is true:
```
>>> 10 ** 2 * 3
300
>>> 10 ** 2 / 3
33.333333333333336
>>> 10 ** 2 + 3
103
>>> 
```

DhruvaG2000 committed:
Move ** operator logic to Nexp func

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

DhruvaG2000 committed:
Update librz/util/calc.c

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

DhruvaG2000 committed:
more tests for ** operator

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

ret2libc commented on librz/util/calc.c:
```diff
@@ -133,6 +133,25 @@ static inline RzNumCalcValue Ndiv(RzNumCalcValue n, RzNumCalcValue v) {
 	return n;
 }
 
+static inline RzNumCalcValue Nexp(RzNumCalcValue n, RzNumCalcValue v) {
```
instead of using `(int)v.d` I *think* you can just use `v.n`, which is essentially the integer representation of the same number.

ret2libc commented on test/db/cmd/cmd_help:
```diff
@@ -214,6 +214,158 @@ trits   0t2
+unit    0
+segment 0000:0000
+string  "\0"
+fvalue  0.1
```
this is wrong?

ret2libc commented on librz/util/calc.c:
```diff
@@ -133,6 +133,25 @@ static inline RzNumCalcValue Ndiv(RzNumCalcValue n, RzNumCalcValue v) {
+static inline RzNumCalcValue Nexp(RzNumCalcValue n, RzNumCalcValue v) {
+	RzNumCalcValue exp_n = n;
+	if (v.d - (int)v.d) {
+		RZ_LOG_WARN("floating point powers not yet supported")
```
```suggestion
		RZ_LOG_WARN("floating point powers not yet supported\n")
```

wargio replied:
add also `;` (semicolumn)

ret2libc commented on librz/util/calc.c:
```diff
@@ -133,6 +133,25 @@ static inline RzNumCalcValue Ndiv(RzNumCalcValue n, RzNumCalcValue v) {
+		for (int i = 1; i < (int)v.d; i++) {
+			n = Nmul(exp_n, n);
+		}
+	} else if ((int)v.d < 0) {
+		for (int i = -1; i >= (int)v.d; i--) {
+			n = Ndiv(n, exp_n);
+		}
```
this seems a bit wrong. Try to play a bit with negative exponents, I see wrong results:
```
? 10**-5
```

To compute the negative I would do `1 / Nexp(n, -exp)` or something equivalent.

DhruvaG2000 committed:
Fix negative power

Adds one more test for negative power and addition and
also implements few other minor changes suggested

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

DhruvaG2000 commented:
I have made the necessary corrections.

On Thu, Jan 20, 2022 at 4:16 PM Giovanni ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In librz/util/calc.c
> <https://github.com/rizinorg/rizin/pull/2220#discussion_r788640127>:
>
> > @@ -133,6 +133,25 @@ static inline RzNumCalcValue Ndiv(RzNumCalcValue n, RzNumCalcValue v) {
>  	return n;
>  }
>
> +static inline RzNumCalcValue Nexp(RzNumCalcValue n, RzNumCalcValue v) {
> +	RzNumCalcValue exp_n = n;
> +	if (v.d - (int)v.d) {
> +		RZ_LOG_WARN("floating point powers not yet supported")
>
> add also ; (semicolumn)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rizinorg/rizin/pull/2220#discussion_r788640127>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJ2S3QDWFJOPQV77PRDXUR3UW7RZFANCNFSM5MHCBEHA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

ret2libc commented on librz/util/calc.c:
```diff
@@ -133,6 +133,25 @@ static inline RzNumCalcValue Ndiv(RzNumCalcValue n, RzNumCalcValue v) {
 
+static inline RzNumCalcValue Nexp(RzNumCalcValue n, RzNumCalcValue v) {
+	RzNumCalcValue exp_n = n;
+	if (v.d - (int)v.n) {
```
use `(st64)` then instead of `(int)`

DhruvaG2000 committed:
librz/util/calc.c: use (st64) then instead of (int)

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

DhruvaG2000 committed:
Merge branch 'dev' into dev-exponent

ret2libc commented on librz/util/calc.c:
```diff
@@ -133,6 +133,25 @@ static inline RzNumCalcValue Ndiv(RzNumCalcValue n, RzNumCalcValue v) {
 }
 
+static inline RzNumCalcValue Nexp(RzNumCalcValue n, RzNumCalcValue v) {
+	RzNumCalcValue exp_n = n;
```
any reason for having this new variable instead of just using `v`?

ret2libc replied:
Ignore this comment, I think I misread it.

ret2libc commented on test/db/cmd/cmd_help:
```diff
@@ -214,6 +214,177 @@ trits   0t2
+unit    0
+segment 0000:0000
+string  "\0"
+fvalue  0.0
```
this is wrong. Please review the code for negative exponents.

DhruvaG2000 replied:
This seems to be an issue concerning the overall code and how it handles floating point. So I am marking as resolved for now.

DhruvaG2000 committed:
Fixes negative exponent bug

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

DhruvaG2000 committed:
Revert "Fixes negative exponent bug"

This was kind of unnecessarily adding many more lines of code while same could be achieved in a single loop

This reverts commit 9657d9bb099610ac01ba0691fc18158461291387.

ret2libc commented on librz/util/calc.c:
```diff
@@ -133,6 +133,25 @@ static inline RzNumCalcValue Ndiv(RzNumCalcValue n, RzNumCalcValue v) {
+		}
+	} else if ((st64)v.n < 0) {
+		for (st64 i = 1; i > (st64)v.n; i--) {
+			n = Ndiv(n, exp_n);
```
I don't understand this... the first time it is executed it does `n / n`, which is ofc 1. I think it should start by doing `1 / n` and go on from there.

ret2libc commented on librz/util/calc.c:
```diff
@@ -133,6 +133,25 @@ static inline RzNumCalcValue Ndiv(RzNumCalcValue n, RzNumCalcValue v) {
+			n = Ndiv(n, exp_n);
+		}
+	} else {
+		n = Ndiv(n, exp_n);
```
same here. Isn't this just the number 1?

ret2libc committed:
Merge branch 'dev' into dev-exponent

XVilka commented:
```


[XX] db/esil/arm_16 cmp eq changed zf
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'e esil.debug=true
e asm.arch=arm
e asm.bits=16
aei
aeim
ar sp=0x00108000
ar r1=0x10
ar cpsr=0x00000000
.arf
wa cmp r1, 16
aes
?v cpsr & 0xf0000000
' malloc://0x200
-- stdout
--- expected
+++ actual
@@ -1,1 +1,1 @@
-0x60000000
+0x5a00000000000000

-- stderr
rz_config_set: variable 'esil.debug' not found



[XX] db/esil/arm_16 cmp eq unchanged zf
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'e esil.debug=true
e asm.arch=arm
e asm.bits=16
aei
aeim
ar sp=0x00108000
ar r1=0x10
ar cpsr=0x40000000
.arf
wa cmp r1, 16
aes
?v cpsr & 0xf0000000
' malloc://0x200
-- stdout
--- expected
+++ actual
@@ -1,1 +1,1 @@
-0x60000000
+0x5a00000000000000

-- stderr
rz_config_set: variable 'esil.debug' not found



[XX] db/esil/arm_16 cmp neq changed zf
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'e esil.debug=true
e asm.arch=arm
e asm.bits=16
aei
aeim
ar sp=0x00108000
ar cpsr=0xffffffff
.arf
wa cmp r1, 16
aes
?v cpsr & 0xf0000000
' malloc://0x200
-- stdout
--- expected
+++ actual
@@ -1,1 +1,1 @@
-0x80000000
+0x86ffffff10000000

-- stderr
rz_config_set: variable 'esil.debug' not found



[XX] db/esil/arm_16 cmp neq unchanged zf
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'e esil.debug=true
e asm.arch=arm
e asm.bits=16
aei
aeim
ar sp=0x00108000
.arf
wa cmp r1, 16
aes
?v cpsr & 0xf0000000
' malloc://0x200
-- stdout
--- expected
+++ actual
@@ -1,1 +1,1 @@
-0x80000000
+0x7800000000000000

-- stderr
rz_config_set: variable 'esil.debug' not found
```

ret2libc committed:
Merge branch 'dev' into dev-exponent

DhruvaG2000 committed:
Merge branch 'rizinorg:dev' into dev-exponent

DhruvaG2000 committed:
** op: add tests for negative numbers

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

DhruvaG2000 committed:
Merge branch 'dev-exponent' of github.com:DhruvaG2000/rizin into dev-exponent

ret2libc committed:
Merge branch 'dev' into dev-exponent

ret2libc commented:
@DhruvaG2000 I'm re-running the CI with updated code. Please monitor the results and if there are failing tests address them, otherwise we can't merge.

DhruvaG2000 commented:
> @DhruvaG2000 I'm re-running the CI with updated code. Please monitor the results and if there are failing tests address them, otherwise we can't merge.

Yes I am monitoring the results, however I fail to understand how my code could be causing any of these errors.
```
[XX] db/esil/arm_32 cmp eq changed zf
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'e esil.debug=true
e asm.arch=arm
e asm.bits=32
aei
aeim
ar sp=0x00108000
ar cpsr=0xffffffff
.arf
wa cmp r1, 16
aes
?v cpsr & 0xf0000000
' malloc://0x200
-- stdout
--- expected
+++ actual
@@ -1,1 +1,1 @@
-0x80000000
+0x86ffffff10000000

-- stderr
rz_config_set: variable 'esil.debug' not found



[XX] db/esil/arm_32 cmp eq unchanged zf
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'e esil.debug=true
e asm.arch=arm
e asm.bits=32
aei
aeim
ar sp=0x00108000
ar r1=0x10
ar cpsr=0x40000000
.arf
wa cmp r1, 16
aes
?v cpsr & 0xf0000000
' malloc://0x200
-- stdout
--- expected
+++ actual
@@ -1,1 +1,1 @@
-0x60000000
+0x5a00000000000000

-- stderr
rz_config_set: variable 'esil.debug' not found



[XX] db/esil/arm_32 cmp neq changed zf
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'e esil.debug=true
e asm.arch=arm
e asm.bits=32
aei
aeim
ar sp=0x00108000
ar cpsr=0xffffffff
.arf
wa cmp r1, 16
aes
?v cpsr & 0xf0000000
' malloc://0x200
-- stdout
--- expected
+++ actual
@@ -1,1 +1,1 @@
-0x80000000
+0x86ffffff10000000

-- stderr
rz_config_set: variable 'esil.debug' not found



[XX] db/esil/arm_32 cmp neq unchanged zf
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'e esil.debug=true
e asm.arch=arm
e asm.bits=32
aei
aeim
ar sp=0x00108000
.arf
wa cmp r1, 16
aes
?v cpsr & 0xf0000000
' malloc://0x200
-- stdout
--- expected
+++ actual
@@ -1,1 +1,1 @@
-0x80000000
+0x7800000000000000

-- stderr
rz_config_set: variable 'esil.debug' not found
```
What exactly is ``esil.debug``?

DhruvaG2000 commented:
> @DhruvaG2000 I'm re-running the CI with updated code. Please monitor the results and if there are failing tests address them, otherwise we can't merge.

Seems like everything was fine until here: https://github.com/rizinorg/rizin/pull/2220/commits/7197ce549cc2e762e67b03ac180ed9d5651b37ce 
Probably something I added after this is causing the issues. not sure if it's some syntax error in the test files or the C source code part.

[update]
The issue was in the c code ``calc.c``. Moving the ``*`` case below did the trick. Let's see (and hope nth else breaks) now.

DhruvaG2000 committed:
Fixes CI test run issues

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

DhruvaG2000 committed:
Merge branch 'dev-exponent' of github.com:DhruvaG2000/rizin into dev-exponent
